// @azure/static-web-apps-cli/dist/core/utils.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mock_fs_1 = __importDefault(require("mock-fs"));
const path_1 = __importDefault(require("path"));
const utils_1 = require("./utils");
describe("Utils", () => {
    beforeEach(() => {
        process.env.DEBUG = "";
        process.argv = [];
    });
    describe("argv()", () => {
        it("process.argv = []", () => {
            process.argv = [];
            expect(utils_1.argv("--port")).toBe(null);
        });
        it("process.argv = ['--port']", () => {
            process.argv = ["--port"];
            expect(utils_1.argv("--port")).toBe(true);
            expect(utils_1.argv("--portxyz")).toBe(false);
        });
        it("process.argv = ['--port=4242']", () => {
            process.argv = ["--port=4242"];
            expect(utils_1.argv("--port")).toBe("4242");
        });
        it("process.argv = ['--port  =   4242  ']", () => {
            process.argv = ["--port  =  4242  "];
            expect(utils_1.argv("--port")).toBe("4242");
        });
        it("process.argv = ['--port', '4242']", () => {
            process.argv = ["--port", "4242"];
            expect(utils_1.argv("--port")).toBe("4242");
        });
        it("process.argv = ['--port', '--other']", () => {
            process.argv = ["--port", "--other"];
            expect(utils_1.argv("--port")).toBe(true);
        });
        it("process.argv = ['--port=']", () => {
            process.argv = ["--port="];
            expect(utils_1.argv("--port")).toBe(null);
        });
    });
    describe("response()", () => {
        it("context = null", () => {
            expect(() => {
                utils_1.response({
                    context: null,
                });
            }).toThrow();
        });
        it("context.bindingData = {foo:bar} (DEBUG off)", () => {
            expect(utils_1.response({
                status: 200,
                context: {
                    bindingData: {
                        foo: "bar",
                    },
                },
            })).toEqual({
                body: null,
                cookies: undefined,
                headers: { "Content-Type": "application/json", status: 200 },
                status: 200,
            });
        });
        it("context.bindingData = {foo:bar} (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {
                        foo: "bar",
                    },
                },
            });
            expect(typeof res.body).toBe("string");
            expect(JSON.parse(res.body).debug).toBeDefined();
            expect(JSON.parse(res.body).debug.context).toBeDefined();
            expect(JSON.parse(res.body).debug.context.foo).toBe("bar");
        });
        it("status = null", () => {
            expect(() => {
                utils_1.response({
                    context: {
                        bindingData: {},
                    },
                    status: null,
                });
            }).toThrow(/TypeError/);
        });
        it("status = 200", () => {
            const res = utils_1.response({
                context: {
                    bindingData: {},
                },
                status: 200,
            });
            expect(res.status).toBe(200);
            expect(res.headers.status).toBe(200);
        });
        it("body = null (DEBUG off)", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                body: null,
            });
            expect(res.body).toBe(null);
        });
        it("body = null (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {
                        foo: "bar",
                    },
                },
                body: null,
            });
            expect(typeof res.body).toBe("string");
            expect(JSON.parse(res.body).debug).toBeDefined();
            expect(JSON.parse(res.body).debug.context).toBeDefined();
            expect(JSON.parse(res.body).debug.context.foo).toBe("bar");
        });
        it("body = {foo:bar} (DEBUG off)", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                body: {
                    foo: "bar",
                },
            });
            expect(typeof res.body).toBe("object");
            expect(res.body.foo).toBeDefined();
            expect(res.body.foo).toBe("bar");
        });
        it("body = {foo:bar} (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                body: {
                    foo: "bar",
                },
            });
            expect(typeof res.body).toBe("object");
            expect(res.body.foo).toBeDefined();
            expect(res.body.foo).toBe("bar");
        });
        it("headers = null (DEBUG off)", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: null,
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = null (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: null,
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = { foo: bar } (DEBUG off)", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: {
                    foo: "bar",
                },
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.foo).toBe("bar");
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = { foo: bar } (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: {
                    foo: "bar",
                },
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.foo).toBe("bar");
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = { location: null } (DEBUG off)", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: {
                    location: null,
                },
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.location).toBe(null);
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = { location: null } (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: {
                    location: null,
                },
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.location).toBe(null);
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = { location: 'wassim.dev' } (DEBUG off)", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: {
                    location: "wassim.dev",
                },
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.location).toBe("wassim.dev");
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("headers = { location: 'wassim.dev' } (DEBUG on)", () => {
            process.env.DEBUG = "*";
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                headers: {
                    location: "wassim.dev",
                },
            });
            expect(res.headers).toBeDefined();
            expect(res.headers.location).toBe(null);
            expect(res.headers.status).toBe(200);
            expect(res.headers["Content-Type"]).toBe("application/json");
        });
        it("cookies = null", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                cookies: null,
            });
            expect(res.cookies).toBe(null);
        });
        it("cookies = { foo:bar }", () => {
            const res = utils_1.response({
                status: 200,
                context: {
                    bindingData: {},
                },
                cookies: {
                    foo: "bar",
                },
            });
            expect(res.cookies).toBeDefined();
            expect(res.cookies.foo).toBe("bar");
        });
    });
    describe("validateCookie()", () => {
        it("cookies = ''", () => {
            expect(utils_1.validateCookie("")).toBe(false);
        });
        it("cookies = 'abc'", () => {
            expect(utils_1.validateCookie("")).toBe(false);
        });
        it("cookies = 'foo=bar'", () => {
            expect(utils_1.validateCookie("foo=bar")).toBe(false);
        });
    });
    describe("readConfigFile()", () => {
        it("config file not found should return undefined", () => {
            expect(utils_1.readConfigFile()).toBe(undefined);
        });
        it("config file with wrong filename should return undefined", () => {
            mock_fs_1.default({
                ".github/workflows/wrong-file-name-pattern.yml": "",
            });
            expect(utils_1.readConfigFile()).toBe(undefined);
            mock_fs_1.default.restore();
        });
        it("invalid YAML file should throw", () => {
            mock_fs_1.default({
                ".github/workflows/azure-static-web-apps__not-valid.yml": "",
            });
            expect(() => utils_1.readConfigFile()).toThrow(/could not parse the SWA workflow file/);
            mock_fs_1.default.restore();
        });
        describe("checking workflow properties", () => {
            it("missing property 'jobs' should throw", () => {
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps__not-valid.yml": `name: Azure Static Web Apps CI/CD`,
                });
                expect(() => utils_1.readConfigFile()).toThrow(/missing property 'jobs'/);
                mock_fs_1.default.restore();
            });
            it("missing property 'jobs.build_and_deploy_job' should throw", () => {
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  invalid_property:
`,
                });
                expect(() => utils_1.readConfigFile()).toThrow(/missing property 'jobs.build_and_deploy_job'/);
                mock_fs_1.default.restore();
            });
            it("missing property 'jobs.build_and_deploy_job.steps' should throw", () => {
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    invalid_property:
`,
                });
                expect(() => utils_1.readConfigFile()).toThrow(/missing property 'jobs.build_and_deploy_job.steps'/);
                mock_fs_1.default.restore();
            });
            it("invalid property 'jobs.build_and_deploy_job.steps' should throw", () => {
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
`,
                });
                expect(() => utils_1.readConfigFile()).toThrow(/missing property 'jobs.build_and_deploy_job.steps'/);
                mock_fs_1.default.restore();
            });
            it("invalid property 'jobs.build_and_deploy_job.steps[]' should throw", () => {
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
`,
                });
                expect(() => utils_1.readConfigFile()).toThrow(/invalid property 'jobs.build_and_deploy_job.steps\[\]'/);
                mock_fs_1.default.restore();
            });
            it("missing property 'jobs.build_and_deploy_job.steps[].with' should throw", () => {
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
`,
                });
                expect(() => utils_1.readConfigFile()).toThrow(/missing property 'jobs.build_and_deploy_job.steps\[\].with'/);
                mock_fs_1.default.restore();
            });
        });
        describe("checking SWA properties", () => {
            it("property 'app_location' should be set", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          app_location: "/"
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.appLocation).toBe(path_1.default.normalize(process.cwd() + "/"));
                mock_fs_1.default.restore();
            });
            it("property 'app_location' should be set to '/' if missing", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          foo: bar
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.appLocation).toBe(path_1.default.normalize(process.cwd() + "/"));
                mock_fs_1.default.restore();
            });
            it("property 'api_location' should be set", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          api_location: "/api"
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.apiLocation).toBe(path_1.default.normalize(process.cwd() + "/api"));
                mock_fs_1.default.restore();
            });
            it("property 'api_location' should be undefined if missing", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          foo: bar
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.apiLocation).toBeUndefined();
                mock_fs_1.default.restore();
            });
            it("property 'app_artifact_location' should be set", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          app_artifact_location: "/"
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.appArtifactLocation).toBe(path_1.default.normalize(process.cwd() + "/"));
                mock_fs_1.default.restore();
            });
            it("property 'app_artifact_location' should be set to '/' if missing", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          foo: bar
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.appArtifactLocation).toBe(path_1.default.normalize(process.cwd() + "/"));
                mock_fs_1.default.restore();
            });
            it("property 'app_build_command' should be set", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          app_build_command: "echo test"
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.appBuildCommand).toBe("echo test");
                mock_fs_1.default.restore();
            });
            it("property 'app_build_command' should be set to default if missing", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          foo: bar
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.appBuildCommand).toBe("npm run build --if-present");
                mock_fs_1.default.restore();
            });
            it("property 'api_build_command' should be set", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          api_build_command: "echo test"
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.apiBuildCommand).toBe("echo test");
                mock_fs_1.default.restore();
            });
            it("property 'api_build_command' should be set to default if missing", () => {
                var _a;
                mock_fs_1.default({
                    ".github/workflows/azure-static-web-apps.yml": `
jobs:
  build_and_deploy_job:
    steps:
      - name: Build And Deploy
        uses: Azure/static-web-apps-deploy@v0.0.1-preview
        with:
          foo: bar
`,
                });
                expect(utils_1.readConfigFile()).toBeTruthy();
                expect((_a = utils_1.readConfigFile()) === null || _a === void 0 ? void 0 : _a.apiBuildCommand).toBe("npm run build --if-present");
                mock_fs_1.default.restore();
            });
        });
    });
});
